File: leetcode/minRefuelStops.js
/**
 * @param {number} target
 * @param {number} startFuel
 * @param {number[][]} stations
 * @return {number}
 */
var minRefuelStops = function (target, startFuel, stations) {
  const dfs = (i, cur, fuel, count) => {
    if (i === stations.length) {
      return cur + fuel >= target ? count : -1;
    }
    if (fuel < 0) return -1;
    const [s, f] = stations[i];
    const d = s - cur;
    if (fuel - d < 0) return -1;
    const skip = dfs(i + 1, s, fuel - d, count);
    if (skip !== -1) return skip;
    return dfs(i + 1, s, fuel - d + f, count + 1);
  };
  return dfs(0, 0, startFuel, 0);
};
minRefuelStops(1000, 299, [
  [13, 21],
  [26, 115],
  [100, 47],
  [225, 99],
  [299, 141],
  [444, 198],
  [608, 190],
  [636, 157],
  [647, 255],
  [841, 123],
]);

File: leetcode/minRefuelStops.py
class Solution(object):
    def minRefuelStops(self, target, startFuel, stations):
        dp = [startFuel] + [0] * len(stations)
        for i, (location, capacity) in enumerate(stations):
            for t in range(i, -1, -1):
                if dp[t] >= location:
                    dp[t + 1] = max(dp[t + 1], dp[t] + capacity)

        for i, d in enumerate(dp):
            if d >= target: return i
        return -1


class Solution(object):
    def minRefuelStops(self, target, tank, stations):
        pq = []  # A maxheap is simulated using negative values
        stations.append((target, float('inf')))

        ans = prev = 0
        for location, capacity in stations:
            tank -= location - prev
            while pq and tank < 0:  # must refuel in past
                tank += -heapq.heappop(pq)
                ans += 1
            if tank < 0: return -1
            heapq.heappush(pq, -capacity)
            prev = location

        return


def minRefuelStops(self, target, cur, s):
    pq = []
    res = i = 0
    while cur < target:
        while i < len(s) and s[i][0] <= cur:
            heapq.heappush(pq, -s[i][1])
            i += 1
        if not pq: return -1
        cur += -heapq.heappop(pq)
        res += 1
    return res